### PR TITLE
implementation of trigger flags with trigger object matching

### PR DESCRIPTION
**Pull request by harrypuuter**
What is still missing is the correct setting of the output shifts. So far, only the base version of the quantitiy is added to the snapshot function, but not the shifted version, and so far, i don't know why exactly. Other than that, the PR is finished.

### CODE REVIEW

**Comment by harrypuuter** [comment]
resolves #59

File: config/config.py [hunk 1]
```diff
@@ -56,6 +57,41 @@ def build_config():
             "tau_ES_shift_DM11": 1.0,
         },
         "mt": {
+            "singlemoun_trigger": [
```
Comment (swozniewski) [inline config/config.py:60]: you could add an era dependency here: "singlemuon_trigger": {"ERA_201...": [...]} and below

Comment (harrypuuter) [reply config/config.py]: Updated

**Review by swozniewski — CHANGES_REQUESTED** [review]

File: analysis_template.cxx [hunk 2]
```diff
@@ -31,7 +32,7 @@ int main(int argc, char *argv[]) {
 
     TStopwatch timer;
     timer.Start();
-
+    ROOT::EnableImplicitMT(1); //
```
[the remark this answers is not in the document]
Comment (harrypuuter) [reply analysis_template.cxx:35]: fixed

File: code_generation/quantity.py [hunk 3]
```diff
@@ -88,6 +90,36 @@ def get_shifts(self, scope):
...
+        quantity.ignored_shifts = self.ignored_shifts
+        quantity.defined_for_scopes = self.defined_for_scopes
+        self.quantities.append(quantity)
+
+    def get_leafs_of_scope(self, scope):
+        # For the writeout, we have to loop over all quantities in the group and return them all (plus their shifts) in a list
+        output = []
+        for quantity in self.quantities:
+            output.extend(
+                [quantity.name]
```
[the remark this answers is not in the document]
Comment (harrypuuter) [reply code_generation/quantity.py]: fixed

File: code_generation/producer.py [hunk 4]
```diff
@@ -196,6 +205,52 @@ def writecalls(self, config, scope):
...
+
+    def writecalls(self, config, scope):
+        n_versions = len(config[""][scope][self.vec_config])
+        log.debug("Number of trigger producers to be created {}".format(n_versions))
+        for i in range(n_versions):
+            self.outputQuantities.add(
+                config[""][scope][self.vec_config][i][self.outputname]
+            )
+            # we also have to replace the quantity in the output scope list
+            config["output"][scope]
```
[the remark this answers is not in the document]
Comment (harrypuuter) [reply code_generation/producer.py]: fixed

File: triggers.hxx [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+            ->debug("bit Value: {}", IntBits(triggerobject_bits[idx]));
+        Logger::get("CheckTriggerMatch")
+            ->debug("bit Value: {}", triggerobject_bits[idx]);
+        auto triggerobject = ROOT::Math::RhoEtaPhiVectorF(
+            0, triggerobject_etas[idx], triggerobject_phis[idx]);
+        // We check the deltaR match as well as that the pt and eta of the
+        // triggerobject are above the given thresholds
+        bool deltaR = ROOT::Math::VectorUtil::DeltaR(triggerobject, particle) <
+                      matchDeltaR;
+        // it we don't want to do any matching here, the triggerbut_cut value is
```
Comment (swozniewski) [inline triggers.hxx:129]: Almost there. it -> if

File: code_generation/producer.py [hunk 6]
```diff
@@ -196,6 +205,52 @@ def writecalls(self, config, scope):
         return calls
 
 
+class TriggerVectorProducer(Producer):
+    def __init__(self, name, call, input, output, scope, vec_config):
+        self.name = name
+        # we create a Quantity Group, which is updated during the writecalls() step
+        self.output = [q.QuantityGroup(name)]
+        self.outputQuantities = self.output[0]
```
[the remark this answers is not in the document]
Comment (harrypuuter) [reply code_generation/producer.py]: changed to an `@property`

**Review by swozniewski — APPROVED** [review]

**Pull request merged**